### PR TITLE
Move hir place

### PR DESCRIPTION
**Pull request by arora-aman**
Needed to support rust-lang/project-rfc-2229#7

This resembles the MIR equivalent located in `lbrustc_middle/mir`.

---
This change is [Reviewable](https://reviewable.io/reviews/sexxi-goose/rust/19)

### CODE REVIEW

**Comment by arora-aman** [comment]
This will let Logan/ Archer/ whoever has some time to work on rust-lang/project-rfc-2229#7 

PR that Chris, Dhruv and I are working on will have some impact, nothing too bad. Hopefully Archer's PR gets merged soon and I will remove the commit from this branch then. 

This builds and that should mostly be it, clippy tests passed. 

Complete test suite running here: http://csclub.uwaterloo.ca/~a52arora/rust-builds/log_commit_861cfd2d0614d8bfeff70c30131a3cacb996c9e1_918a4302-c93d-4474-8291-c3f48d787aae

**Comment by arora-aman** [comment]
Discussion relating to move the HIR Place to middle/hir: https://rust-lang.zulipchat.com/#narrow/stream/189812-t-compiler.2Fwg-rfc-2229/topic/Typecheck.20tables.20using.20Places/near/201773438

**Review by Azhng — APPROVED** [review]
Looks fairly mechanical. LGTM

File: src/librustc_middle/hir/place.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+pub enum ProjectionKind {
+    /// A dereference of a pointer, reference or `Box<T>` of the given type
+    Deref,
+    /// An index or a field
+    Other,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable, HashStable)]
+pub struct Projection<'tcx> {
+    // Type after the projection is being applied.
```
Comment (arora-aman) [inline src/librustc_middle/hir/place.rs]: Make this three /// so that docs are generated

File: src/librustc_middle/hir/place.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        self.projections.iter().enumerate().rev().filter_map(move |(index, proj)| {
+            if ProjectionKind::Deref == proj.kind {
+                Some(self.ty_before_projection(index))
+            } else {
+                None
+            }
+        })
+    }
+
+    // Returns the type of this `Place` after all projections have been applied.
```
Comment (arora-aman) [inline src/librustc_middle/hir/place.rs]: three ///

File: src/librustc_middle/hir/place.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    pub fn ty(&self) -> Ty<'tcx> {
+        self.projections.last().map_or_else(|| self.base_ty, |proj| proj.ty)
+    }
+
+    // Returns the type of this `Place` immediately before `projection_index`th projection
```
Comment (arora-aman) [inline src/librustc_middle/hir/place.rs]: three

**Review by arora-aman — COMMENTED** [review]
Fix docs

**Comment by arora-aman** [comment]
Rebased and tests passed: http://csclub.uwaterloo.ca/~a52arora/rust-builds/log_06bef156-2ef2-4836-bb78-b06b0d17dde7

**Comment by arora-aman** [comment]
rust-lang PR: https://github.com/rust-lang/rust/pull/74424

**Pull request merged**